File: backend/app/schemas/storage.py
from pydantic import BaseModel, ConfigDict, Field

from app.models.storage import RackType
from app.schemas.units import MaterialUnitOut


class WarehouseCreate(BaseModel):
    name: str
    address: str | None = None


class WarehouseOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: int
    name: str
    address: str | None
    is_active: bool


class WarehouseUpdate(BaseModel):
    name: str | None = None
    address: str | None = None
    is_active: bool | None = None


class RackCreate(BaseModel):
    code: str
    type: RackType
    shelf_count: int = Field(gt=0)
    strip_capacity: int | None = Field(default=None, gt=0)
    warehouse_id: int


class RackOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: int
    code: str
    type: RackType
    shelf_count: int
    strip_capacity: int | None
    warehouse_id: int
    is_active: bool


class RackUpdate(BaseModel):
    code: str | None = None
    type: RackType | None = None
    shelf_count: int | None = Field(default=None, gt=0)
    strip_capacity: int | None = Field(default=None, gt=0)
    warehouse_id: int | None = None
    is_active: bool | None = None


class MacroZoneRuleCreate(BaseModel):
    from_shelf: int = Field(gt=0)
    to_shelf: int = Field(gt=0)
    material: str | None = None
    color: str | None = None
    thickness: float | None = None
    manufacturer: str | None = None


class MacroZoneRuleOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: int
    rack_id: int
    from_shelf: int
    to_shelf: int
    material_id: int | None
    color_id: int | None
    thickness_id: int | None
    manufacturer_id: int | None


class LocationSuggestion(BaseModel):
    location_code: str | None


class RackOccupancyCellOut(BaseModel):
    """Одна полка стеллажа (4.1 ТЗ) — рулонные и штрипсовые адресуются
    одинаково (полка целиком, без ячеек). `capacity` — сколько единиц
    может занимать этот адрес одновременно (1 для рулонных, где полка =
    один рулон; `Rack.strip_capacity` для штрипсовых). `units` — что
    сейчас физически на полке, пустой список = свободна."""

    shelf: int
    location_code: str
    units: list[MaterialUnitOut]
    capacity: int

File: frontend/src/api/storage.ts
import { apiClient } from "./client";
import type { MaterialUnit } from "./units";

export type RackType = "roll" | "strip";

export interface Warehouse {
  id: number;
  name: string;
  address: string | null;
  is_active: boolean;
}

export interface WarehouseCreate {
  name: string;
  address?: string;
}

export interface WarehouseUpdate {
  name?: string;
  address?: string;
  is_active?: boolean;
}

export interface Rack {
  id: number;
  code: string;
  type: RackType;
  shelf_count: number;
  warehouse_id: number;
  is_active: boolean;
}

export interface RackUpdate {
  code?: string;
  type?: RackType;
  shelf_count?: number;
  warehouse_id?: number;
  is_active?: boolean;
}

export const listWarehouses = async (): Promise<Warehouse[]> =>
  (await apiClient.get<Warehouse[]>("/warehouses")).data;

export const createWarehouse = async (payload: WarehouseCreate): Promise<Warehouse> =>
  (await apiClient.post<Warehouse>("/warehouses", payload)).data;

export const updateWarehouse = async (warehouseId: number, payload: WarehouseUpdate): Promise<Warehouse> =>
  (await apiClient.patch<Warehouse>(`/warehouses/${warehouseId}`, payload)).data;

export interface MacroZoneRule {
  id: number;
  rack_id: number;
  from_shelf: number;
  to_shelf: number;
  material_id: number | null;
  color_id: number | null;
  thickness_id: number | null;
  manufacturer_id: number | null;
}

export interface MacroZoneRuleCreate {
  from_shelf: number;
  to_shelf: number;
  material?: string;
  color?: string;
  thickness?: number;
  manufacturer?: string;
}

export const listRacks = async (warehouseId?: number): Promise<Rack[]> =>
  (await apiClient.get<Rack[]>("/racks", { params: warehouseId ? { warehouse_id: warehouseId } : undefined })).data;

export const createRack = async (payload: {
  code: string;
  type: RackType;
  shelf_count: number;
  warehouse_id: number;
}): Promise<Rack> => (await apiClient.post<Rack>("/racks", payload)).data;

export const updateRack = async (rackId: number, payload: RackUpdate): Promise<Rack> =>
  (await apiClient.patch<Rack>(`/racks/${rackId}`, payload)).data;

export const listMacroZoneRules = async (rackId: number): Promise<MacroZoneRule[]> =>
  (await apiClient.get<MacroZoneRule[]>(`/racks/${rackId}/macro-zone-rules`)).data;

export const createMacroZoneRule = async (rackId: number, payload: MacroZoneRuleCreate): Promise<MacroZoneRule> =>
  (await apiClient.post<MacroZoneRule>(`/racks/${rackId}/macro-zone-rules`, payload)).data;

export const deleteMacroZoneRule = async (rackId: number, ruleId: number): Promise<void> => {
  await apiClient.delete(`/racks/${rackId}/macro-zone-rules/${ruleId}`);
};

export interface RackOccupancyCell {
  shelf: number;
  cell: number | null;
  location_code: string;
  unit: MaterialUnit | null;
}

export const getRackOccupancy = async (rackId: number): Promise<RackOccupancyCell[]> =>
  (await apiClient.get<RackOccupancyCell[]>(`/racks/${rackId}/occupancy`)).data;

export const suggestLocation = async (params: {
  material_sku_id: number;
  width_mm: number;
  parent_id?: number | null;
  warehouse_id?: number;
}): Promise<string | null> =>
  (await apiClient.get<{ location_code: string | null }>("/storage/suggest-location", { params })).data
    .location_code;
